### ISSUE TITLE
Allow HSM adaptive moments to use circular weight function

### ISSUE THREAD

**Issue by rearmstr** [GalSim-developers/GalSim#917]
The HSM adaptive moments code that is exposed uses an elliptical weight function when computing moments.  There is also existing code in the HSM source file to use a circular weight function.  This issue is to modify the call to the adaptive moments code to allow for the use of a circular weight function.

**Comment by rearmstr** [comment]
@rmandelb, I think the code is working now, but now I'm not sure what is the best radius to return.  

The ```find_mom_2``` function returns the width of the best fit Gaussian and a set of moments.  My first thought was to use the returned width, but doing some measurements on simple sheared Gaussians, this number seems to give the same results as using the elliptical weights.  Maybe I am doing something wrong here?

**Comment by rmandelb** [comment]
Are you comparing the `moments_sigma` outputs for the two cases?  If so, then I think what you're finding is what we expect.  The `moments_sigma` output is based on the determinant so it should be relatively insensitive to shear.  The issue we were having with the EstimateShear resolution factor is because it is based on the trace (which is what we want for noisy images because it's linear in the moments, but it has this problematic shape dependence).  

So the idea (I thought) was that you'd have the new code return the `moments_sigma` and `observed_shape` from the circularly-weighted moments, and we'd use that to get a trace estimate that is less sensitive to the shear.  But the determinant which is in `moments_sigma` is already insensitive to shape for a Gaussian model.

**Comment by rmandelb** [comment]
BTW I tried using your code, but was getting exceptions.  Is the latest version pushed?

**Comment by rearmstr** [comment]
> Are you comparing the moments_sigma outputs for the two cases? If so, then I think what you're finding is what we expect. The moments_sigma output is based on the determinant so it should be relatively insensitive to shear. The issue we were having with the EstimateShear resolution factor is because it is based on the trace (which is what we want for noisy images because it's linear in the moments, but it has this problematic shape dependence).

Yes, that is what I was comparing.  I see what you mean.  How would you recommend returning the observed shape?  There is a moments matrix that is calculated, but I'm not sure what is computed exactly.

> BTW I tried using your code, but was getting exceptions. Is the latest version pushed?
Strange.  Yes, pushed version should be the latest.

**Comment by rmandelb** [comment]
> There is a moments matrix that is calculated, but I'm not sure what is computed exactly.

It should simply be the moments weighted by the round weight function, so you can do the usual thing of taking the 2nd moment matrix elements and using it to calculate e1 and e2, which go into the `observed_e1` and `observed_e2` fields in the ShapeData object that FindAdaptiveMomView returns.

As for the exceptions, here's a code snippet that reproduces the issue:
```
import galsim

pix_scale = 0.1
sigma_vals = [0.5,0.7]
g1_vals = [0.0, 0.1, 0.5]
g2_vals = [0.0, 0.2]

def make_gauss_im(sigma, g1, g2, scale):
    g = galsim.Gaussian(sigma=sigma)
    if g1>0 or g2>0:
        g = g.shear(g1=g1, g2=g2)
    im = g.drawImage(scale=scale, method='no_pixel')
    return im

for sigma in sigma_vals:
    for g1 in g1_vals:
        for g2 in g2_vals:
            im = make_gauss_im(sigma, g1, g2, pix_scale)
            res1 = im.FindAdaptiveMom()
            res2 = im.FindAdaptiveMom(round_moments=True)
            print sigma, g1, g2, res1.moments_sigma, res1.observed_shape, res2.moments_sigma, res2.observed_shape
```
The first time it gets to that line that calls FindAdaptiveMom with `round_moments=True`, it dies with this error:
```
RuntimeError: TMV Error: Failed Assert statement m2.colsize() == colsize() && m2.rowsize() == rowsize()
```
Does that code work for you?

**Comment by rearmstr** [comment]
> It should simply be the moments weighted by the round weight function, so you can do the usual thing of taking the 2nd moment matrix elements and using it to calculate e1 and e2, which go into the observed_e1 and observed_e2 fields in the ShapeData object that FindAdaptiveMomView returns.

Hmm, my question was how do you get the xx and yy moments from the matrix that is returned.  Sorry if I'm being totally dense here.

> Does that code work for you?

Yes, it runs with no problems.

**Comment by rmandelb** [comment]
Oh, I totally misunderstood your question!  Sorry.  The TMV moments stuff was put in place by Mike, so I am less conversant with how to use them.  Reading later in src/hsm/PSFCorr.cpp about how these are defined, there's a comment that suggests that the right formula is
```
e1=0.5*(moments(2,0)-moments(0,2))/moments(0,0)
e2=(1/sqrt(2)) moments(1,1)/moments(0,0)
```
This is not what I would have guessed.

When I tried this, the results were not insane, in the sense that (a) if I apply g1 or g2 of some magnitude, then the weighted g1 and g2 estimates are consistent with each other. (b) the weighted g1 and g2 estimates are rounder than the applied shear.

**Comment by rmandelb** [comment]
And BTW I found that the script I posted earlier crashed on my laptop, which has TMV 0.74, and runs fine on the cluster here, which has TMV 0.72.  What version of TMV are you using?

**Comment by rmjarvis** [comment]
> The TMV moments stuff was put in place by Mike, so I am less conversant with how to use them.
I'm pretty sure I just converted the existing code to use TMV.  I didn't set up how the moments matrix works.  ([Here](https://github.com/GalSim-developers/GalSim/blob/e42701ed110f2b4d91c031a3de81ea845d2c7d3b/src/hsm/PSFCorr.cpp) is a very old version you can compare to, which has the same kind of sqrt(2) stuff.)

Also, I ran the code you posted above on a machine with TMV v0.74, and it worked fine for me.  Here was the output:
```
0.5 0.0 0.0 5.0 galsim.Shear(g1=-2.84217087951e-16,g2=-6.34107889469e-18) 5.0 galsim.Shear(g1=0.0,g2=0.0)
0.5 0.0 0.2 5.0 galsim.Shear(g1=-8.18545206049e-16,g2=0.200000003874) 5.0 galsim.Shear(g1=0.0,g2=0.0)
0.5 0.1 0.0 5.0 galsim.Shear(g1=0.100000000304,g2=6.01828049942e-18) 5.0 galsim.Shear(g1=0.0,g2=0.0)
0.5 0.1 0.2 5.0 galsim.Shear(g1=0.100000002059,g2=0.200000004117) 5.0 galsim.Shear(g1=0.0,g2=0.0)
0.5 0.5 0.0 5.0 galsim.Shear(g1=0.500000012418,g2=1.00136428339e-17) 5.0 galsim.Shear(g1=0.0,g2=0.0)
0.5 0.5 0.2 5.0 galsim.Shear(g1=0.500000012995,g2=0.200000001354) 5.0 galsim.Shear(g1=0.0,g2=0.0)
0.7 0.0 0.0 7.0 galsim.Shear(g1=0.0,g2=7.41026236238e-19) 7.0 galsim.Shear(g1=0.0,g2=0.0)
0.7 0.0 0.2 7.0 galsim.Shear(g1=-6.26437680163e-16,g2=0.200000003874) 7.0 galsim.Shear(g1=0.0,g2=0.0)
0.7 0.1 0.0 7.0 galsim.Shear(g1=0.100000000304,g2=-6.59740525521e-18) 7.0 galsim.Shear(g1=0.0,g2=0.0)
0.7 0.1 0.2 7.0 galsim.Shear(g1=0.100000002059,g2=0.200000004117) 7.0 galsim.Shear(g1=0.0,g2=0.0)
0.7 0.5 0.0 7.0 galsim.Shear(g1=0.500000012418,g2=-4.128763178e-19) 6.99999904633 galsim.Shear(g1=0.0,g2=0.0)
0.7 0.5 0.2 7.0 galsim.Shear(g1=0.500000012995,g2=0.200000001354) 6.99999856949 galsim.Shear(g1=0.0,g2=0.0)
```

**Comment by rmandelb** [comment]
@rmjarvis - That is really strange.  I have no idea what could be going on (with my TMV error).

@rearmstr - in case helpful, I sat down this morning and tried to remind myself what the "right" answer is for an elliptical Gaussian with an adaptive circular Gaussian weight function, so you can use that in code tests.  If you start with an elliptical Gaussian that has `sigma_x = sqrt(q) sigma` and `sigma_y = sigma / sqrt(q)` (so that the axis ratio is q), then with elliptical adaptive weights you should get the usual `|e| = (1-q^2) / (1+q^2)` when defining `e` using the `(Mxx-Myy)/(Mxx+Myy)` and `2 Mxy/(Mxx+Myy)` definition of shape.  If you make the weight function a circular Gaussian with that same `sigma` as above (which would happen when doing adaptive circular moments) then I believe you should get `|e| = (1-q^2) / (1 + q^2 + 2q)`.  So it's nearly the same answer except for that extra 2q in the denominator, .

I think that if you make circular Gaussians and shear them with axis ratio `q`, you should be able to use the above formula to test that the code I suggested above is correctly giving you e1 and e2.

**Comment by rearmstr** [comment]
> And BTW I found that the script I posted earlier crashed on my laptop, which has TMV 0.74, and runs fine on the cluster here, which has TMV 0.72. What version of TMV are you using?

I'm using TMV 0.73.

**Comment by rmjarvis** [comment]
> That is really strange. I have no idea what could be going on (with my TMV error).

I figured that much out.  If I compile GalSim with TMV_DEBUG=true, then I get the same error running that code.  Which does mean there is probably a bug somewhere, since that's testing that the sizes of matrix operations on the left and right side of an assignment are compatible.  I'll check it out and try to give a more concrete prescription of what the problem is.

**Comment by rearmstr** [comment]
I think I found the problem.  When calling ```find_mom_2```, the argument to max_order needs to be one less than the size of the matrix.  Before I had it as the same.

@rmandelb, I tried to compare the formula you gave for ```e``` with those that I measured via the code.  I found that I get the right answer if I multiply the measured moments by ```sqrt(2)```, ie: 

e1=sqrt(2)*0.5*(moments(2,0)-moments(0,2))/moments(0,0)
e2=moments(1,1)/moments(0,0)
